lib/models/database/post_repository.dart

```diff
@@ -0,0 +1,278 @@
+  static const String geohashField = "geohash"; // Do not change
+
+  const PostLocationFirestore({
+    required this.geoPoint,
+    required this.geohash,
+  });
+
+  /// Parses the data from a firestore document
+  factory PostLocationFirestore.fromDbData(Map<String, dynamic> data) {
+    try {
+      return PostLocationFirestore(
+        geoPoint: data[geoPointField],
+        geohash: data[geohashField],
+      );
+    } catch (e) {
+      if (e is TypeError) {
+        throw Exception("Cannot parse post location document: ${e.toString()}");
+      } else {
+        rethrow;
+      }
+    }
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other)) return true;
+
+    return other is PostLocationFirestore &&
+        other.geoPoint == geoPoint &&
+        other.geohash == geohash;
+  }
+
+  @override
+  int get hashCode => Object.hash(geoPoint, geohash);
+}
+
+class PostRepository {
+  final GeoFlutterFire _geoFire;
+  final GeoLocationService _geoLocationService;
+
+  static const collectionName = "posts";
+  final CollectionReference _collectionRef;
+
+  static const double kmPostRadius = 0.1;
+
+  PostRepository({
+    required FirebaseFirestore firestore,
+    required GeoFlutterFire geoFire,
+    required GeoLocationService geoLocationService,
+  })  : _collectionRef = firestore.collection(collectionName),
+        _geoFire = geoFire,
+        _geoLocationService = geoLocationService;
+
+  /// Get the posts near the current location of the user
+  /// The radius is defined by [kmPostRadius]
+  Future<List<PostFirestore>> getNearPosts() async {
+    final point = await _getCurrentGeoFirePoint();
+
+    return _getPostsNearLocation(point, kmPostRadius);
+  }
+
+  /// Get a post by its id
+  Future<PostFirestore> getPost(PostFirestoreId postId) async {
+    final docSnap = await _collectionRef.doc(postId).get();
+
+    return PostFirestore.fromDb(docSnap);
+  }
+
+  /// Adds a post at the current location of the user
+  Future<void> addPost(PostFirestoreData postData) async {
+    final point = await _getCurrentGeoFirePoint();
+
+    return _addPostAtLocation(point, postData);
+  }
+
+  /// Deletes a post by its id
+  Future<void> deletePost(PostFirestoreId postId) async {
+    await _collectionRef.doc(postId).delete();
+  }
+
+  Future<void> _addPostAtLocation(
+    GeoFirePoint point,
+    PostFirestoreData postData,
+  ) async {
+    // The `point.data` returns a Map<String, dynamic> consistent with the
+    // class [PostLocationFirestore]. This is because the field name values
+    // are hardcoded in the [GeoFlutterFire] library
+
+    await _collectionRef.add({
+      PostFirestore.locationField: point.data,
+      ...postData.toDbData(),
+    });
+  }
+
+  Future<List<PostFirestore>> _getPostsNearLocation(
+    GeoFirePoint point,
+    double radius,
+  ) async {
+    final posts = await _geoFire
+        .collection(collectionRef: _collectionRef)
+        .within(
```

Comment:
I am not sure at all of this comment. In [the documentation](https://pub.dev/packages/geoflutterfire2), they say that `within` may cause "memory leak, because underlying stream listeners are not cancelled when you cancel StreamSubscription from within method", and that `withinAsSingleStreamSubscription` is much safer. I was wondering if you thought about this and `within` was a concious choice? 
I really struggle to understand the (too small) documentation of this package, but, since the stream is cancelled imediately after on the `first`, don't we have a memory leak?

Reply:
You're right, I have stubble accross this issue and tried to found an alternative but somehow I completely missed the `withinAsSingleStreamSubscription` method (I don't really know how).

So thanks for pointing it out and I fix it in commit 0b1ed553ca519807cfb50389256f908ee725c2f5.